pkg/controller/v1alpha1/localmodel/controller.go

```diff
@@ -680,3 +686,75 @@ func (c *LocalModelReconciler) getContainerSpecForStorageUri(storageUri string)
 	}
 	return defaultContainer, nil
 }
+
+// UpdateLocalModelNode updates the source model uri of the localmodelnode from the localmodel
+func (c *LocalModelReconciler) UpdateLocalModelNode(ctx context.Context, localmodelNode *v1alpha1.LocalModelNode, localModel *v1alpha1api.ClusterLocalModel, jobNamespace string) error {
+	var patch client.Patch
+	updated := false
+	for i, modelInfo := range localmodelNode.Spec.LocalModels {
+		if modelInfo.ModelName == localModel.Name {
+			if modelInfo.SourceModelUri == localModel.Spec.SourceModelUri {
+				return nil
+			}
+			// Update the source model uri
+			updated = true
+			patch = client.MergeFrom(localmodelNode.DeepCopy())
+			localmodelNode.Spec.LocalModels[i].SourceModelUri = localModel.Spec.SourceModelUri
+			break
+		}
+	}
+	if !updated {
+		patch = client.MergeFrom(localmodelNode.DeepCopy())
+		localmodelNode.Spec.LocalModels = append(localmodelNode.Spec.LocalModels, v1alpha1api.LocalModelInfo{ModelName: localModel.Name, SourceModelUri: localModel.Spec.SourceModelUri})
+	}
+	if err := c.Client.Patch(context.TODO(), localmodelNode, patch); err != nil {
+		c.Log.Error(err, "Update localmodelnode", "name", localmodelNode.Name)
+		return err
+	}
+	return nil
+}
+
+// ReconcileLocalModelNode creates updates localmodelnode for each node in the node group. It adds and removes localmodels from the localmodelnode and updates the status on the localmodel from the localmodelnode.
+func (c *LocalModelReconciler) ReconcileLocalModelNode(ctx context.Context, localModel *v1alpha1api.ClusterLocalModel, nodeGroup *v1alpha1api.LocalModelNodeGroup, jobNamespace string) error {
+	readyNodes, notReadyNodes, err := getNodesFromNodeGroup(nodeGroup, c.Client)
+	if err != nil {
+		c.Log.Error(err, "getNodesFromNodeGroup node error")
+		return err
+	}
+	for _, node := range notReadyNodes.Items {
+		if _, ok := localModel.Status.NodeStatus[node.Name]; !ok {
+			localModel.Status.NodeStatus[node.Name] = v1alpha1api.NodeNotReady
+		}
+	}
+	for _, node := range readyNodes.Items {
+		localModelNode := &v1alpha1.LocalModelNode{}
+		err := c.Client.Get(ctx, types.NamespacedName{Name: node.Name}, localModelNode)
+		found := true
+		if err != nil {
+			if apierr.IsNotFound(err) {
+				found = false
+				c.Log.Info("localmodelNode not found")
+			} else {
+				c.Log.Error(err, "Failed to get localmodelnode", "name", node.Name)
+				return err
+			}
+		}
+		if !found {
+			localModelNode = &v1alpha1.LocalModelNode{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: node.Name,
+				},
+				Spec: v1alpha1api.LocalModelNodeSpec{LocalModels: []v1alpha1api.LocalModelInfo{{ModelName: localModel.Name, SourceModelUri: localModel.Spec.SourceModelUri}}},
+			}
+			if err := c.Client.Create(ctx, localModelNode); err != nil {
```

Comment:
Does local model node agent waits for this resource to be created before starting reconciliation ?

Reply:
Yes!